### PR TITLE
Added driver for MLX90614 temperature sensor

### PR DESCRIPTION
**Pull request by rahulOCR**
Hey everyone, 
I am submitting this PR to merge into master. I have added MLX90614 temperature sensor into AP_TemperatureSensor library.
Using datas-sheet provided [Link of data-sheet]](https://www.melexis.com/en/documents/documentation/datasheets/datasheet-mlx90614). I have used this to get ambient temperature of cabin of battery.

### CODE REVIEW

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor.h [hunk 1]
```diff
@@ -37,6 +38,7 @@ class AP_TemperatureSensor
     friend class AP_TemperatureSensor_TSYS03;
     friend class AP_TemperatureSensor_Analog;
     friend class AP_TemperatureSensor_DroneCAN;
+    class AP_TemperatureSensor_MLX90614;
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor.h:41]:
```suggestion
    friend class AP_TemperatureSensor_MLX90614;
```

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    // RAM reg - Linearized temperature, source #1
+    // _crude_value = read_data(MLX90614_TOBJ1);
+    
+    
+    if(_crude_value > 0)
+    {
+        // debugg
+        // printf("MLX90614: temp = %f\n",(_crude_value * 0.02 - 273.15));
+        // temp * 0.02 - 273.15 = degrees, temp * 0.02 is temperature in kelvin
+        set_temperature((_crude_value * 0.02 - 273.15));
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp:61]:
```suggestion
        WITH_SEMAPHORE(_dev->get_semaphore());
        set_temperature((_crude_value * 0.02 - 273.15));
```

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+#include "AP_TemperatureSensor_MLX90614.h"
+
+
+extern const AP_HAL::HAL &hal;
+
+
+void AP_TemperatureSensor_MLX90614::init()
+{
+    constexpr char name[] = "MLX90614";
+    (void)name;  // sometimes this is unused (e.g. HAL_GCS_ENABLED false)
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp:15]: this shouldn't be necessary.

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    if ((_params.bus_address != MLX90614_I2CDEFAULTADDR) ) {
+        printf("%s wrong I2C addr of 0x%2X, setting to 0x%2X", name, (unsigned)_params.bus_address.get(), (unsigned)MLX90614_I2CDEFAULTADDR);
+        _params.bus_address.set(MLX90614_I2CDEFAULTADDR);
+    }
+    
+    // _dev = std::move(hal.i2c_mgr->get_device(_params.bus, _params.bus_address));
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp:22]: please remove unused/commented code

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        // printf("MLX90614: temp = %f\n",(_crude_value * 0.02 - 273.15));
+        // temp * 0.02 - 273.15 = degrees, temp * 0.02 is temperature in kelvin
+        set_temperature((_crude_value * 0.02 - 273.15));
+    }
+    
+
+}
+
+
+void AP_TemperatureSensor_MLX90614::update()
```
Comment (magicrub) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp:68]: remove, you've already declared it empty in the header

**Review by magicrub — CHANGES_REQUESTED** [review]

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp (hunk 2 above)
Comment (magicrub) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp:61]: for clarity, do this conversion computation on a separate line above it and store it into a const float

**Comment by magicrub** [comment]
please rebase onto master and squash your commits

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp (hunk 5 above)
Comment (rahulOCR) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp]: done

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp (hunk 4 above)
Comment (rahulOCR) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp]: resolved

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp (hunk 3 above)
Comment (rahulOCR) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp]: resolved

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp (hunk 2 above)
Comment (rahulOCR) [reply libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp]: resolved

**Comment by rahulOCR** [comment]
rebase done

**Comment by rahulOCR** [comment]
Hi I am attaching datasheet, wiki and photos for interfacing. MLX90614 sensor is an infrared thermometer for non-contact temperature measurements. Both the IR sensitive thermopile detector chip and the signal conditioning ASIC are integrated in the same TO-39 can. Integrated into the MLX90614 are a low noise amplifier, 17-bit ADC and powerful DSP unit thus achieving high accuracy and resolution of the thermometer.

set (examples shown for first sensor):

[TEMP1_TYPE](https://ardupilot.org/copter/docs/parameters.html#temp1-type) = 0:Disabled, 1:TSYS01, 2:MCP9600, 3:MAX31865, 4:TSYS03, 5:Analog, 6:DroneCAN, 7:MLX90614 Sensor type

[TEMP1_BUS](https://ardupilot.org/copter/docs/parameters.html#temp1-bus) = 0-3 Board’s I2C port on which the sensor is attached

[TEMP1_ADDR](https://ardupilot.org/copter/docs/parameters.html#temp1-addr) = the sensor’s I2C address (0-127). For this sensor, 90 

[TEMP1_SRC](https://ardupilot.org/copter/docs/parameters.html#temp1-src) = which other temperature reporting device type should have its temperature report replaced with this sensor’s.

[TEMP1_SRC_ID](https://ardupilot.org/copter/docs/parameters.html#temp1-src-id) = this identifies which individual device, of the above type, should have its report replaced.
   
 
[
wiki_MLX90614

[MLX90614-Datasheet-Melexis.pdf](https://github.com/ArduPilot/ardupilot/files/14823016/MLX90614-Datasheet-Melexis.pdf)

**Pull request closed without merging**